Reduce max length of slug fields to 752 characters

jasujm opened the pull request:
Recent versions of mysql limit the size of key length to 3072 bytes, effectively limiting the size of varchar indices to 767 characters when using utf8mb4 encoding. An earlier PR (https://github.com/nephila/djangocms-blog/pull/419) sets the minimum length of slug fields to 767 characters to address the issue. However, when slug fields appear in composite indices, the limit of 767 characters applies to the _composite_ index and not just the slug part. This limits the maximum length even further.

With freshly installed database (I used the official mysql:8.0.19 docker image for testing) running migrations leads to this:

```
$ python manage.py migrate
  Apply all migrations: ...
Running migrations:
  Applying contenttypes.0001_initial... OK
  ...
  Applying djangocms_blog.0032_auto_20180109_0023...Traceback (most recent call last):
  ...
  django.db.utils.OperationalError: (1071, 'Specified key was too long; max key length is 3072 bytes')
```

This is due to slug field in `Post` and `BlogCategory` models appearing in composite index with `language_code`. `language_code` has type varchar(15), effectively limiting the maximum lenght of `slug` to 752 characters (752+15=767) to ensure successful migrations.

This PR suggests to reduce maximum length of slug fields to 752 characters. It was done to existing migrations. While this may cause inconsistencies for old installations that have already applied those migrations, it's necessary for fresh installations using utf8mb4 to apply the migrations in the first place.

**Edit:** I added more details after submitting the original PR.

jasujm commented:
Note since the initial PR I pushed an update to the pull request. I left `title` and `subtitle` in the `Post` model intact. Those do not appear in composite indices, so it's not necessary to edit them.

yakky commented on djangocms_blog/models.py:
```diff
@@ -89,7 +89,7 @@ class BlogCategory(BlogMetaMixin, TranslatableModel):
 
     translations = TranslatedFields(
         name=models.CharField(_('name'), max_length=767),
```
`name` field must also be reduced, or the slug truncated when generated, otherwise you may get errors when saving a name longer than the allowed slug field length

jasujm replied:
I shortened it. Also shortened the `title` field in blog post, as post slug is derived from it by default.

yakky commented on djangocms_blog/models.py:
```diff
@@ -225,7 +225,7 @@ class Post(KnockerModel, BlogMetaMixin, TranslatableModel):
     translations = TranslatedFields(
         title=models.CharField(_('title'), max_length=767),
-        slug=models.SlugField(_('slug'), max_length=767, blank=True,
+        slug=models.SlugField(_('slug'), max_length=752, blank=True,
```
`title` field must also be reduced, or the slug truncated when generated, otherwise you may get errors when saving a name longer than the allowed slug field length

jasujm commented:
I pushed a new version with the following changes:
- Reduced the size of `title` and `subtitle` in the Post model and `name` in the Category model. I understand reducing the length of `subtitle` is a matter of taste and maybe unnecessary, but I felt it's justified for consistency.
- Added new migration applying the changes also for existing installations. Note that the retroactive edits in old migrations is still necessary for new installations.
- Added a summary in HISTORY.rst about what I changed

yakky commented:
@jasujm I would leave `subtitle` out of the change, as it's not touched by the issue and I would rather apply the least changes possible
Other than that, I tried to break :smiling_imp: the PR but it looks :ok_hand: even with pre-existing data
This will need a rebase after merge of #542 which fixes unrelated issues with python 2

jasujm commented:
Thanks! I removed the changes I made to `subtitle`. I also rebased on top of the head of the `develop` branch.

yakky commented:
@jasujm thanks a lot for your contribution

codecompagnons commented:
@yakky We tried all this but we still get the "django.db.utils.OperationalError: (1071, 'Specified key was too long; max key length is 3072 bytes')" error. What more could be wrong? 
We are running:
python 3.8
Django cms 3.7.1
Django 2.2
on Pythonanywhere.

I hope there is an solution..

Thanks in advance!
Johan

yakky commented:
which version of mysql are you using?

codecompagnons commented:
> which version of mysql are you using?

Hi, thanks for your reply. MySQL 5.6.48

yakky commented:
@codecompagnons thanks I will try to reproduce the issue
